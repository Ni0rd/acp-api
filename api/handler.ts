import { IncomingMessage, ServerResponse } from 'http';
import { ApolloServer } from 'apollo-server-micro';
import microCors from 'micro-cors';
import jsonwebtoken from 'jsonwebtoken';
import typeDefs from './typeDefs';
import resolvers from './resolvers';

if (!process.env.JWT_SECRET) {
  throw new Error('JWT_SECRET missing');
}
if (!process.env.WP_API_ENDPOINT) {
  throw new Error('WP_API_ENDPOINT missing');
}

interface Context {
  req: IncomingMessage;
  res: ServerResponse;
  token?: {};
}

const apolloServer = new ApolloServer({
  typeDefs,
  resolvers,
  context(ctx: Context): Context {
    const { authorization } = ctx.req.headers;
    const token = authorization ? authorization.split(' ').pop() : null;
    ctx.token = {};
    if (token) {
      try {
        ctx.token = jsonwebtoken.verify(
          token,
          process.env.JWT_SECRET as string
        );
      } catch (err) {
        // Ignore errors
      }
    }
    return ctx;
  },
  introspection: true,
  playground: true,
});

// Apollo cannot handle OPTIONS requests, ignore them
// @see: https://github.com/apollographql/apollo-server/issues/2473
const corsHandler = (req: IncomingMessage, res: ServerResponse): void => {
  if (req.method === 'OPTIONS') {
    res.end();
    return;
  }
  apolloServer.createHandler()(req, res);
};

export default microCors()(corsHandler);
