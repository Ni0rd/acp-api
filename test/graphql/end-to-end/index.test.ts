import './env';
import { Server } from 'http';
import fetch from 'cross-fetch';
import gql from 'graphql-tag';
import { ApolloClient } from 'apollo-client';
import { InMemoryCache, NormalizedCacheObject } from 'apollo-cache-inmemory';
import { HttpLink } from 'apollo-link-http';
import { setupServer } from './helpers';
import handler from '../../../api/graphql';

let server!: Server;
let handlerUrl!: string;
let client: ApolloClient<NormalizedCacheObject>;

beforeEach(async () => {
  const result = await setupServer(handler);
  server = result.server;
  handlerUrl = result.url;

  const cache = new InMemoryCache();
  const link = new HttpLink({
    fetch,
    uri: `${handlerUrl}/graphql`,
  });
  client = new ApolloClient({
    cache,
    link,
  });
});

afterEach(() => {
  server.close();
});

describe('GraphQL Handler', () => {
  test('Returns the plan categories', async () => {
    const response = await client.query({
      query: gql`
        query($lang: Lang!) {
          planCategories(lang: $lang) {
            id
          }
        }
      `,
      variables: {
        lang: 'fr',
      },
    });
    expect(response).toMatchSnapshot();
  });
});
