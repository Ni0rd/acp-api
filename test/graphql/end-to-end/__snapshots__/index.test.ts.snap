// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GraphQL Handler Returns the plan categories 1`] = `
Object {
  "data": Object {
    "planCategories": Array [
      Object {
        "__typename": "PlanCategory",
        "id": "1",
      },
      Object {
        "__typename": "PlanCategory",
        "id": "2",
      },
      Object {
        "__typename": "PlanCategory",
        "id": "3",
      },
      Object {
        "__typename": "PlanCategory",
        "id": "4",
      },
      Object {
        "__typename": "PlanCategory",
        "id": "5",
      },
    ],
  },
  "loading": false,
  "networkStatus": 7,
  "stale": false,
}
`;
